### PR TITLE
Eleventy build pipeline image resizing, conversion and retina handling

### PR DESCRIPTION
**Pull request by Pezmc**
**Merge target is**: https://github.com/flowforge/website/pull/458 do not merge before that is in main

## Description

This PR significantly reduces the size of images served on the website by converting them into modern web formats like WebP and ensuring images have the correct resolution.

Key advantages:
 - Team do not need to consider image sizing and file format
 - Faster page load times
 - Improved SEO ranking

Key disadvantages:
 - New syntax to recal
 - Slower _first build_ (about 20 seconds) as all images are generated

It does this by introducing two new image pipelines taking advantage of  to the blog:
 - One that adds a shortcode `{% image %}` for use in Nunjucks files
 - Another that handles all images included in markdown files

Will provide full usage details as part of https://github.com/flowforge/handbook/issues/220, but in summary:

For markdown, any image included in the markup is automatically downsized to 650px wide (and if the image is large enough 1300px for retina screens); and converted to webp, avif and jpeg. Gifs are converted to webp only.

For nunjucks, there's a new shortcode {% image, that should be used instead of <img:

```nunjucks
{% image "path-to-image.jpg", "Required ALT tag" %}
```

Ideally, the maximum width the image will be displayed on the page should also be passed to resize the image:

```nunjucks
{% image "path-to-image.jpg", "Required ALT tag". [450] %}
```

The above will automatically take into account retina screens and generate both a 450px and 900px wide image. The users browser will select which to display.

The new files are copied to `/img`, for backwards compatibility (e.g. CSS files) "old images" are still copied to `/images` too.

## Related Issue(s)

#401 #349

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [x] I have considered the performance impact of these changes
 - [-] Suitable unit/system level tests have been added and they pass
 - [-] Documentation has been updated - https://github.com/flowforge/handbook/issues/220

### CODE REVIEW

File: src/webinars.njk [hunk 1]
```diff
@@ -130,12 +134,19 @@ meta:
                     <a href="{{ item.url }}" class="w-full flex flex-col group hover:no-underline">
                         <div class="">
                             <time class="block text-xs mb-2" value="{{ item.date }}" class="text-gray-400">{{ item.date | shortDate }}</time>
-                            <div>{% if item.image %}<img class="w-full h-auto" src="{{item.image}}" />{% else %}{{ item.url | generatePostSVG |safe}}{% endif %}</div>
```
Comment (Pezmc) [inline src/webinars.njk:133]: `item.image` appears to be a bug

Comment (knolleary) [reply src/webinars.njk:133]: Why a bug? Isn't that intended to add an image if the front matter defines one?

Comment (joepavitt) [reply src/webinars.njk:133]: this is fixed in https://github.com/flowforge/website/pull/460

Comment (joepavitt) [reply src/webinars.njk:133]: > Why a bug? Isn't that intended to add an image if the front matter defines one?

it used to work, but when we upgraded to Eleventy 1.0, I think I missed this one as we didn't have any past webinars in that flagged it.

**Comment by joepavitt** [comment]
unable to `npm install` in a fresh instal of the repo from this branch unfortunately. It's an odd one:

```
website % npm install
npm ERR! code 1
npm ERR! path /Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/sharp
npm ERR! command failed
npm ERR! command sh -c (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)
npm ERR! sharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.13.3/libvips-8.13.3-darwin-arm64v8.tar.br
npm ERR! node:internal/fs/utils:344
npm ERR!     throw err;
npm ERR!     ^
npm ERR! 
npm ERR! Error: EACCES: permission denied, copyfile '/var/folders/mk/8rcqb0t12w77fvl_4kd7bngh0000gn/T/38678-libvips-8.13.3-darwin-arm64v8.tar.br' -> '/Users/joepavitt/.npm/_libvips/libvips-8.13.3-darwin-arm64v8.tar.br'
npm ERR!     at Object.copyFileSync (node:fs:2800:3)
npm ERR!     at WriteStream.<anonymous> (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/sharp/install/libvips.js:205:20)
npm ERR!     at WriteStream.emit (node:events:390:28)
npm ERR!     at emitCloseNT (node:internal/streams/destroy:138:10)
npm ERR!     at processTicksAndRejections (node:internal/process/task_queues:82:21) {
npm ERR!   errno: -13,
npm ERR!   syscall: 'copyfile',
npm ERR!   code: 'EACCES',
npm ERR!   path: '/var/folders/mk/8rcqb0t12w77fvl_4kd7bngh0000gn/T/38678-libvips-8.13.3-darwin-arm64v8.tar.br',
npm ERR!   dest: '/Users/joepavitt/.npm/_libvips/libvips-8.13.3-darwin-arm64v8.tar.br'
npm ERR! }

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/joepavitt/.npm/_logs/2023-02-28T08_54_53_728Z-debug.log```

**Comment by joepavitt** [comment]
seems related to https://stackoverflow.com/questions/54409953/cant-install-sharp

not comfortable with the solution, any idea where this has been introduced @Pezmc?

**Comment by joepavitt** [comment]
just switched back to `main` to fresh install there, and it's definitely something on this branch causing the issue

**Comment by Pezmc** [comment]
Sharp is part of the [image processing pipeline](https://www.npmjs.com/package/sharp). It depends on libvips to handle the processing. You can use `npm why` e.g. `npm why sharp` to look up where a dependency comes from:

```bash
$ npm why sharp
sharp@0.31.3 dev
node_modules/sharp
  sharp@"^0.31.3" from @11ty/eleventy-img@3.0.0
  node_modules/@11ty/eleventy-img
    dev @11ty/eleventy-img@"^3.0.0" from the root project
```

The `npm install --unsafe-perm` flag is ok if you're happy to npm to be able to read the /var/ directory while it handles the first install. It seems related to issues on windows with install scripts. There's a blog post linked from that stackoverflow answer that gives more context: https://geedew.com/What-does-unsafe-perm-in-npm-actually-do/

**Comment by joepavitt** [comment]
yeah, just actually tried `npm install --unsafe-perm` and it's still throwing the same error (on an M1 Mac btw)

**Comment by joepavitt** [comment]
> You can use npm why e.g. npm why sharp to look up where a dependency comes from

TIL

**Comment by Pezmc** [comment]
@joepavitt's `~/.npm` directory had a copy of libvips that was owned by root, rather than his user. The solution was to chown the directory recursively to ensure his user was able to write to `~/.npm/libvips`.

`sudo chown -R $USER ~/.npm`

It's likely that libvips was previously using sudo (from another package) and had left the folders owned by root as a result.

**Comment by joepavitt** [comment]
can now `npm install`, but hitting a new issue:

```
[11ty] Problem writing Eleventy templates: (more in DEBUG output)
[11ty] 1. Having trouble rendering liquid template ./src/docs/user/migration.md (via TemplateContentRenderError)
[11ty] 2. `statsSync` is not supported with remote sources. Use `statsByDimensionsSync` instead. (via Error)
[11ty] 
[11ty] Original error stack trace: Error: `statsSync` is not supported with remote sources. Use `statsByDimensionsSync` instead.
[11ty]     at Function.statsSync (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/@11ty/eleventy-img/img.js:559:13)
[11ty]     at Object.markdownLib.renderer.rules.image (/Users/joepavitt/Documents/flowforge/development/flowforge/website/.eleventy.js:415:40)
[11ty]     at Renderer.renderInline (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/markdown-it/lib/renderer.js:275:28)
[11ty]     at Renderer.render (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/markdown-it/lib/renderer.js:330:22)
[11ty]     at MarkdownIt.render (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/markdown-it/lib/index.js:544:24)
[11ty]     at /Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/@11ty/eleventy/src/Engines/Markdown.js:79:38
[11ty]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[11ty]     at async Template._render (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/@11ty/eleventy/src/TemplateContent.js:507:22)
[11ty]     at async Template.renderWithoutLayout (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/@11ty/eleventy/src/Template.js:456:27)
[11ty]     at async TemplateMap.populateContentDataInMap (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/@11ty/eleventy/src/TemplateMap.js:585:39)
[11ty] Copied 214 files / Wrote 0 files in 34.00 seconds (v2.0.0)
[11ty] Unhandled rejection in promise: (more in DEBUG output)
[11ty] Bad response for https://lh5.googleusercontent.com/rY4vl2YoHINNMoK0MTTrLITrBJgr9Detxd6w3KRJQWXkvRoEv6625Ope0QG3WmCTtRHeJFOk6KCoBfniONIjSP1lxNTltEROkpys8j0CsHuPqKM74HoNMqt0AYufg8z9JRDsrJOa197op8hXwZzgiDU (403): Forbidden (via Error)
[11ty] 
[11ty] Original error stack trace: Error: Bad response for https://lh5.googleusercontent.com/rY4vl2YoHINNMoK0MTTrLITrBJgr9Detxd6w3KRJQWXkvRoEv6625Ope0QG3WmCTtRHeJFOk6KCoBfniONIjSP1lxNTltEROkpys8j0CsHuPqKM74HoNMqt0AYufg8z9JRDsrJOa197op8hXwZzgiDU (403): Forbidden
[11ty]     at RemoteAssetCache.fetch (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/@11ty/eleventy-fetch/src/RemoteAssetCache.js:73:11)
[11ty]     at processTicksAndRejections (node:internal/process/task_queues:96:5)
[11ty]     at async run (/Users/joepavitt/Documents/flowforge/development/flowforge/website/node_modules/p-queue/dist/index.js:163:29)
ERROR: "dev:eleventy" exited with 1.
```

**Comment by joepavitt** [comment]
had to delete the contents in my `/website/src/docs` manually, and then restart the website in order to get clean docs build.

**Review by joepavitt — APPROVED** [review]
Just added a note in the readme to inform users that the `start` command can be a little slow first time round, as it can feel like it's stuck/frozen as it parses the hundreds of images we have.

**Pull request merged**